File: ui/web/src/api/schemas/news.ts
import {z} from 'zod';

import {isIsoDate} from '@/lib/time';

/** The backend counts characters (Python `len`), not UTF-16 units. */
export const EXCERPT_MAX_CHARS = 280;

const isoDateSchema = z.string().refine(isIsoDate, 'Expected YYYY-MM-DD');
// UTC with a `Z`, as the contract says (no offsets).
const utcDateTimeSchema = z.iso.datetime();

export const runStatusSchema = z.enum(['RUNNING', 'DONE', 'FAILED']);

/**
 * A rule engine reason code. Later increments add codes, so any code of this
 * shape is accepted; the UI shows an unknown one as a neutral badge.
 */
export const REASON_CODE_PATTERN = /^[A-Z][A-Z_]{1,40}$/;
export const reasonCodeSchema = z.string().regex(REASON_CODE_PATTERN);

/** How the rules matched a duplicate to the story it copies. */
export const dupTypeSchema = z.enum(['url', 'exact', 'near', 'paraphrase']);

/** Which rule check wrote a piece of evidence. */
export const ruleCheckSchema = z.enum(['entity', 'source', 'dedup', 'stale']);

/** The AI summary's tone about the story (not advice about the security). */
export const sentimentSchema = z.enum(['bullish', 'neutral', 'bearish']);

/** Which AI-run step wrote a piece of evidence. */
export const aiCheckSchema = z.enum(['guard', 'language', 'dedup']);

/** What the AI run did with an item. */
export const aiStatusSchema = z.enum([
  'DONE',
  'SKIPPED',
  'DUPLICATE',
  'FAILED',
]);

/** One line of at most 280 characters (the backend counts code points). */
const oneLineSummarySchema = z
  .string()
  .refine(
    text => [...text].length <= EXCERPT_MAX_CHARS,
    `At most ${EXCERPT_MAX_CHARS} characters`,
  )
  .refine(text => !/[\r\n]/.test(text), 'One line only');

/** Wire format of the ingest run of a date. */
export const ingestRunWireSchema = z.object({
  run_id: z.number().int(),
  status: runStatusSchema,
  started_at: utcDateTimeSchema,
  // null while the run is RUNNING (legacy.ingest_run.finished_at).
  finished_at: utcDateTimeSchema.nullable(),
  rows_received: z.number().int().nonnegative(),
  dups: z.number().int().nonnegative(),
});

/** Wire format of one item in `GET /news`. */
export const newsItemWireSchema = z.object({
  id: z.number().int(),
  vendor_item_id: z.string().min(1),
  feed_date: isoDateSchema,
  headline: z.string(),
  excerpt: z
    .string()
    .refine(
      text => [...text].length <= EXCERPT_MAX_CHARS,
      `At most ${EXCERPT_MAX_CHARS} characters`,
    ),
  source_url: z.string(),
  source_domain: z.string(),
  published_at: utcDateTimeSchema,
  tickers: z.array(z.string()),
  synthetic: z.boolean(),
  // From the rule engine when `rules_checked`, else the legacy exact hash.
  is_dup: z.boolean(),
  dup_of: z.string().nullable(),
  // Sorted by the backend; the UI orders the badges itself.
  reason_codes: z.array(reasonCodeSchema),
  dup_type: dupTypeSchema.nullable(),
  copies: z.number().int().nonnegative(),
  rules_checked: z.boolean(),
  // Null until the AI run summarized the item (increment 3).
  summary: oneLineSummarySchema.nullable(),
  sentiment: sentimentSchema.nullable(),
});

/** Wire format of one rule check's explanation (plain text). */
export const ruleEvidenceWireSchema = z.object({
  check: ruleCheckSchema,
  code: reasonCodeSchema.nullable(),
  message: z.string(),
});

/** Wire format of one AI-run finding (plain text). */
export const aiEvidenceWireSchema = z.object({
  check: aiCheckSchema,
  code: reasonCodeSchema.nullable(),
  message: z.string(),
});

/** A company the model found in the item. */
export const aiCompanyWireSchema = z.object({
  name: z.string(),
  ticker: z.string().nullable(),
});

/** Wire format of what the AI run did with one item. */
export const aiDetailWireSchema = z.object({
  status: aiStatusSchema,
  model: z.string(),
  prompt_version: z.string(),
  enriched_at: utcDateTimeSchema,
  // `fallback`: the model couldn't summarize; the lead sentence is shown.
  summary_source: z.enum(['llm', 'fallback']).nullable(),
  companies: z.array(aiCompanyWireSchema),
  claims: z.array(z.string()),
  evidence: z.array(aiEvidenceWireSchema),
});

/** Wire format of `GET /news/{id}`: the item plus its full body. */
export const newsDetailWireSchema = newsItemWireSchema.extend({
  body: z.string(),
  rule_evidence: z.array(ruleEvidenceWireSchema),
  ai: aiDetailWireSchema.nullable(),
});

/** Wire format of the latest rule-check run of a date. */
export const ruleRunWireSchema = z.object({
  status: runStatusSchema,
  finished_at: utcDateTimeSchema.nullable(),
  items: z.number().int().nonnegative(),
  duplicates: z.number().int().nonnegative(),
  flagged: z.number().int().nonnegative(),
});

/** Wire format of the latest AI run (summaries, paraphrases) of a date. */
export const aiRunWireSchema = z.object({
  status: runStatusSchema,
  finished_at: utcDateTimeSchema.nullable(),
  items: z.number().int().nonnegative(),
  paraphrases: z.number().int().nonnegative(),
  conflicts: z.number().int().nonnegative(),
  summarized: z.number().int().nonnegative(),
  fallbacks: z.number().int().nonnegative(),
  failed: z.number().int().nonnegative(),
  model: z.string(),
});

/** Wire format of `GET /news`. */
export const newsListWireSchema = z
  .object({
    date: isoDateSchema,
    run: ingestRunWireSchema.nullable(),
    rule_run: ruleRunWireSchema.nullable(),
    ai_run: aiRunWireSchema.nullable(),
    count: z.number().int().nonnegative(),
    items: z.array(newsItemWireSchema),
  })
  .refine(list => list.count === list.items.length, {
    message: 'count must equal the number of items',
    path: ['count'],
  });

export type NewsItemWire = z.infer<typeof newsItemWireSchema>;
export type IngestRunWire = z.infer<typeof ingestRunWireSchema>;
export type RuleRunWire = z.infer<typeof ruleRunWireSchema>;
export type RuleEvidenceWire = z.infer<typeof ruleEvidenceWireSchema>;
export type AiRunWire = z.infer<typeof aiRunWireSchema>;
export type AiDetailWire = z.infer<typeof aiDetailWireSchema>;
export type AiEvidenceWire = z.infer<typeof aiEvidenceWireSchema>;

function toNewsItem(wire: NewsItemWire) {
  return {
    id: wire.id,
    vendorItemId: wire.vendor_item_id,
    feedDate: wire.feed_date,
    headline: wire.headline,
    excerpt: wire.excerpt,
    sourceUrl: wire.source_url,
    sourceDomain: wire.source_domain,
    publishedAt: wire.published_at,
    tickers: wire.tickers,
    synthetic: wire.synthetic,
    isDup: wire.is_dup,
    dupOf: wire.dup_of,
    reasonCodes: wire.reason_codes,
    dupType: wire.dup_type,
    copies: wire.copies,
    rulesChecked: wire.rules_checked,
    summary: wire.summary,
    sentiment: wire.sentiment,
  };
}

function toAiRun(wire: AiRunWire) {
  return {
    status: wire.status,
    finishedAt: wire.finished_at,
    items: wire.items,
    paraphrases: wire.paraphrases,
    conflicts: wire.conflicts,
    summarized: wire.summarized,
    fallbacks: wire.fallbacks,
    failed: wire.failed,
    model: wire.model,
  };
}

function toAiDetail(wire: AiDetailWire) {
  return {
    status: wire.status,
    model: wire.model,
    promptVersion: wire.prompt_version,
    enrichedAt: wire.enriched_at,
    summarySource: wire.summary_source,
    // Same shapes and names in the UI model.
    companies: wire.companies,
    claims: wire.claims,
    evidence: wire.evidence,
  };
}

function toIngestRun(wire: IngestRunWire) {
  return {
    runId: wire.run_id,
    status: wire.status,
    startedAt: wire.started_at,
    finishedAt: wire.finished_at,
    rowsReceived: wire.rows_received,
    dups: wire.dups,
  };
}

function toRuleRun(wire: RuleRunWire) {
  return {
    status: wire.status,
    finishedAt: wire.finished_at,
    items: wire.items,
    duplicates: wire.duplicates,
    flagged: wire.flagged,
  };
}

/** `GET /news` as the UI uses it. */
export const newsListSchema = newsListWireSchema.transform(wire => ({
  date: wire.date,
  run: wire.run === null ? null : toIngestRun(wire.run),
  ruleRun: wire.rule_run === null ? null : toRuleRun(wire.rule_run),
  aiRun: wire.ai_run === null ? null : toAiRun(wire.ai_run),
  count: wire.count,
  items: wire.items.map(toNewsItem),
}));

/** `GET /news/{id}` as the UI uses it. */
export const newsDetailSchema = newsDetailWireSchema.transform(wire => ({
  ...toNewsItem(wire),
  body: wire.body,
  // Same shape and names in the UI model.
  ruleEvidence: wire.rule_evidence,
  ai: wire.ai === null ? null : toAiDetail(wire.ai),
}));

export type RunStatus = z.infer<typeof runStatusSchema>;
export type NewsList = z.output<typeof newsListSchema>;
export type NewsItem = NewsList['items'][number];
export type IngestRun = NonNullable<NewsList['run']>;
export type RuleRun = NonNullable<NewsList['ruleRun']>;
export type AiRun = NonNullable<NewsList['aiRun']>;
export type Sentiment = z.infer<typeof sentimentSchema>;
export type AiDetail = NonNullable<NewsDetail['ai']>;
export type AiEvidence = z.infer<typeof aiEvidenceWireSchema>;
export type DupType = z.infer<typeof dupTypeSchema>;
export type RuleEvidence = z.infer<typeof ruleEvidenceWireSchema>;
export type NewsDetail = z.output<typeof newsDetailSchema>;
export type NewsListWire = z.infer<typeof newsListWireSchema>;
export type NewsDetailWire = z.infer<typeof newsDetailWireSchema>;

File: python/ai-api/src/ai_api/news.py
"""The news feed: raw news through the ``ai.v_*`` views, plus rule results.

From increment 2 the API reads raw news only through ``ai.v_raw_news`` and
``ai.v_ingest_run`` (the strangler-fig seam; they switch from ``legacy.*`` to
``ingest.*`` in increment 6), joined with the rule engine's results.

Duplicates: once the rules have checked an item, ``is_dup`` / ``dup_of``
come from the rule engine (URL, exact and near copies, also of earlier
days). Before that, they're the legacy exact-hash flags.

Verdicts (increment 4): a unique item shows its own; a duplicate shows its
original's (``verdict_source = inherited``), except a stale copy, which is
re-served old news and so MISLEADING.
"""

from __future__ import annotations

import dataclasses
import datetime
import re
from typing import Any, Protocol

from psycopg import rows
from psycopg_pool import AsyncConnectionPool

EXCERPT_MAX_CHARS = 280
# A day is about 100 items; this only guards against a runaway feed.
_MAX_ITEMS = 1000
_WHITESPACE = re.compile(r"\s+")

_RUN_SQL = """
SELECT run_id, status, started_at, finished_at, rows_received, dups
FROM ai.v_ingest_run
WHERE feed_date = %(day)s
ORDER BY started_at DESC, run_id DESC
LIMIT 1
"""

_RULE_RUN_SQL = """
SELECT status, finished_at, items, duplicates, flagged
FROM ai.rule_run
WHERE feed_date = %(day)s
ORDER BY started_at DESC, run_id DESC
LIMIT 1
"""

# One row per raw item, with the rule results when the rules have run.
_ITEMS_CTE = """
WITH items AS (
  SELECT r.id, r.vendor_item_id, r.feed_date, r.headline, r.body,
         r.source_url, r.source_domain, r.published_at, r.tickers,
         r.synthetic,
         (c.news_id IS NOT NULL) AS rules_checked,
         CASE WHEN c.news_id IS NULL THEN r.is_dup
              ELSE d.news_id IS NOT NULL END AS is_dup,
         CASE WHEN c.news_id IS NULL THEN r.dup_of
              ELSE o.vendor_item_id END AS dup_of,
         d.dup_type,
         -- Rule codes, then the AI run's (guard, language, L3), then the
         -- verification's, no repeats.
         coalesce(c.reason_codes, '{}') || ARRAY(
           SELECT code FROM unnest(coalesce(a.reason_codes, '{}')) AS code
           WHERE code <> ALL (coalesce(c.reason_codes, '{}'))
         ) || ARRAY(
           SELECT code FROM unnest(coalesce(v.reason_codes, '{}')) AS code
           WHERE code <> ALL (coalesce(c.reason_codes, '{}')
                              || coalesce(a.reason_codes, '{}'))
         ) AS reason_codes,
         c.evidence,
         a.status AS ai_status, a.summary, a.sentiment, a.summary_source,
         a.evidence AS ai_evidence, a.model AS ai_model,
         a.prompt_version, a.enriched_at,
         n.id AS news_id,
         CASE WHEN v.news_id IS NOT NULL THEN v.verdict::text
              WHEN d.news_id IS NOT NULL AND d.stale THEN 'MISLEADING'
              ELSE cv.verdict::text END AS verdict,
         CASE WHEN v.news_id IS NOT NULL THEN v.confidence
              WHEN d.news_id IS NOT NULL AND d.stale THEN 0.9
              ELSE cv.confidence END AS confidence,
         CASE WHEN v.news_id IS NOT NULL THEN v.review_status
              ELSE cv.review_status END AS review_status,
         CASE WHEN v.news_id IS NOT NULL AND v.verdict IS NOT NULL
                THEN 'ai'
              WHEN d.news_id IS NOT NULL AND (d.stale OR cv.verdict IS NOT NULL)
                THEN 'inherited' END AS verdict_source,
         -- Copies in the same feed (a later day's stale copy isn't here).
         (SELECT count(*) FROM ai.duplicate_link x
          JOIN ai.news_item xn ON xn.id = x.news_id
          WHERE x.canonical_id = n.id
            AND xn.feed_date = n.feed_date) AS copies
  FROM ai.v_raw_news r
  LEFT JOIN ai.news_item n
    ON n.feed_date = r.feed_date AND n.vendor_item_id = r.vendor_item_id
  LEFT JOIN ai.rule_check c ON c.news_id = n.id
  LEFT JOIN ai.duplicate_link d ON d.news_id = n.id
  LEFT JOIN ai.news_item o ON o.id = d.canonical_id
  LEFT JOIN ai.news_ai a ON a.news_id = n.id
  LEFT JOIN ai.verification v ON v.news_id = n.id
  LEFT JOIN ai.verification cv ON cv.news_id = d.canonical_id
  WHERE {where}
)
"""

_LIST_SQL = (
    _ITEMS_CTE.replace("{where}", "r.feed_date = %(day)s")
    + f"""
SELECT * FROM items
WHERE (%(ticker)s::text IS NULL OR %(ticker)s::text = ANY (tickers))
  AND (%(dups)s OR NOT is_dup)
  AND (%(pattern)s::text IS NULL
       OR headline ILIKE %(pattern)s OR body ILIKE %(pattern)s)
  AND (%(verdict)s::text IS NULL OR verdict = %(verdict)s)
  AND (NOT %(pending)s OR review_status = 'PENDING')
ORDER BY published_at DESC, id DESC
LIMIT {_MAX_ITEMS}
"""
)

_AI_RUN_SQL = """
SELECT status, finished_at, items, paraphrases, conflicts, summarized,
       fallbacks, failed, model
FROM ai.ai_run
WHERE feed_date = %(day)s
ORDER BY started_at DESC, run_id DESC
LIMIT 1
"""

_EXTRACTION_SQL = """
SELECT 'entity' AS kind, e.seq, e.name AS text, e.ticker
FROM ai.v_raw_news r
JOIN ai.news_item n USING (feed_date, vendor_item_id)
JOIN ai.entity e ON e.news_id = n.id
WHERE r.id = %(id)s
UNION ALL
SELECT 'claim', c.seq, c.text, NULL
FROM ai.v_raw_news r
JOIN ai.news_item n USING (feed_date, vendor_item_id)
JOIN ai.claim c ON c.news_id = n.id
WHERE r.id = %(id)s
ORDER BY 1, 2
"""

_DETAIL_SQL = (
    _ITEMS_CTE.replace("{where}", "r.id = %(id)s") + "SELECT * FROM items"
)

_VERIFY_RUN_SQL = """
SELECT * FROM ai.verify_run
WHERE feed_date = %(day)s
ORDER BY requested_at DESC, run_id DESC
LIMIT 1
"""

# The item's own verification, or its original's for a duplicate.
_VERIFICATION_SQL = """
SELECT v.*
FROM ai.v_raw_news r
JOIN ai.news_item n USING (feed_date, vendor_item_id)
LEFT JOIN ai.duplicate_link d ON d.news_id = n.id
JOIN ai.verification v ON v.news_id = coalesce(d.canonical_id, n.id)
WHERE r.id = %(id)s
"""

_EVIDENCE_SQL = """
SELECT e.seq, e.check_type AS "check", e.code, e.message, e.source, e.url,
       e.title
FROM ai.evidence e
WHERE e.news_id = %(news_id)s
ORDER BY e.seq
"""

_REVIEW_SQL = """
SELECT id, status, reasons, ai_verdict, final_verdict, reviewer, comment,
       created_at, decided_at
FROM ai.review_task
WHERE news_id = %(news_id)s
ORDER BY created_at DESC, id DESC
LIMIT 1
"""


@dataclasses.dataclass(frozen=True)
class NewsQuery:
    """Filters of ``GET /news``.

    Attributes:
        day: The feed date.
        ticker: Only items tagged with this ticker (upper case).
        text: Case-insensitive substring of the headline or body.
        include_duplicates: Also return items flagged as duplicates.
        verdict: Only items with this verdict.
        pending_review: Only items waiting for an analyst.
    """

    day: datetime.date
    ticker: str | None = None
    text: str | None = None
    include_duplicates: bool = False
    verdict: str | None = None
    pending_review: bool = False


class NewsStore(Protocol):
    """Reads ingest runs, rule runs and news items."""

    async def latest_run(self, day: datetime.date) -> dict[str, Any] | None:
        """Returns the latest ingest run of ``day``, or None."""
        ...

    async def latest_rule_run(
        self, day: datetime.date
    ) -> dict[str, Any] | None:
        """Returns the latest rule run of ``day``, or None."""
        ...

    async def latest_ai_run(self, day: datetime.date) -> dict[str, Any] | None:
        """Returns the latest AI run of ``day``, or None."""
        ...

    async def items(self, query: NewsQuery) -> list[dict[str, Any]]:
        """Returns the matching items, newest first (with ``body``)."""
        ...

    async def item(self, item_id: int) -> dict[str, Any] | None:
        """Returns one item (with ``body`` and ``evidence``), or None."""
        ...

    async def extraction(self, item_id: int) -> list[dict[str, Any]]:
        """The item's extracted companies and claims (``kind``, ``text``)."""
        ...

    async def latest_verify_run(
        self, day: datetime.date
    ) -> dict[str, Any] | None:
        """Returns the latest verify run of ``day``, or None."""
        ...

    async def verification(
        self, item_id: int
    ) -> tuple[dict[str, Any], list[dict[str, Any]], dict[str, Any] | None]:
        """The verification, its evidence and the latest review task.

        A duplicate gets its original's. ``({}, [], None)`` when the item
        isn't verified.
        """
        ...


def like_pattern(text: str) -> str:
    r"""Turns user text into an ILIKE pattern that matches it literally.

    Args:
        text: The search text.

    Returns:
        ``%text%`` with ``\``, ``%`` and ``_`` escaped.
    """
    escaped = text.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
    return f"%{escaped}%"


def excerpt(body: str) -> str:
    """The first characters of a body, as one line of plain text.

    Args:
        body: The full body (untrusted vendor text; returned as text, never
            interpreted).

    Returns:
        At most 280 characters (code points, like Python's ``len``), with
        whitespace collapsed and an ellipsis when cut.
    """
    text = _WHITESPACE.sub(" ", body).strip()
    if len(text) <= EXCERPT_MAX_CHARS:
        return text
    return text[: EXCERPT_MAX_CHARS - 1].rstrip() + "…"


class PostgresNewsStore:
    """NewsStore over the ``ai`` views and rule tables."""

    def __init__(self, pool: AsyncConnectionPool) -> None:
        """Uses connections from ``pool``."""
        self._pool = pool

    async def _one(
        self, sql: str, params: dict[str, Any]
    ) -> dict[str, Any] | None:
        async with self._pool.connection() as conn:
            cur = conn.cursor(row_factory=rows.dict_row)
            await cur.execute(sql, params)
            return await cur.fetchone()

    async def latest_run(self, day: datetime.date) -> dict[str, Any] | None:
        """Returns the latest ingest run of ``day``, or None."""
        return await self._one(_RUN_SQL, {"day": day})

    async def latest_rule_run(
        self, day: datetime.date
    ) -> dict[str, Any] | None:
        """Returns the latest rule run of ``day``, or None."""
        return await self._one(_RULE_RUN_SQL, {"day": day})

    async def latest_ai_run(self, day: datetime.date) -> dict[str, Any] | None:
        """Returns the latest AI run of ``day``, or None."""
        return await self._one(_AI_RUN_SQL, {"day": day})

    async def items(self, query: NewsQuery) -> list[dict[str, Any]]:
        """Returns the matching items, newest first (with ``body``)."""
        params = {
            "day": query.day,
            "ticker": query.ticker,
            "dups": query.include_duplicates,
            "pattern": None if query.text is None else like_pattern(query.text),
            "verdict": query.verdict,
            "pending": query.pending_review,
        }
        async with self._pool.connection() as conn:
            cur = conn.cursor(row_factory=rows.dict_row)
            await cur.execute(_LIST_SQL, params)
            return await cur.fetchall()

    async def item(self, item_id: int) -> dict[str, Any] | None:
        """Returns one item (with ``body`` and ``evidence``), or None."""
        return await self._one(_DETAIL_SQL, {"id": item_id})

    async def extraction(self, item_id: int) -> list[dict[str, Any]]:
        """The item's extracted companies and claims (``kind``, ``text``)."""
        async with self._pool.connection() as conn:
            cur = conn.cursor(row_factory=rows.dict_row)
            await cur.execute(_EXTRACTION_SQL, {"id": item_id})
            return await cur.fetchall()

    async def latest_verify_run(
        self, day: datetime.date
    ) -> dict[str, Any] | None:
        """Returns the latest verify run of ``day``, or None."""
        return await self._one(_VERIFY_RUN_SQL, {"day": day})

    async def verification(
        self, item_id: int
    ) -> tuple[dict[str, Any], list[dict[str, Any]], dict[str, Any] | None]:
        """See ``NewsStore.verification``."""
        found = await self._one(_VERIFICATION_SQL, {"id": item_id})
        if found is None:
            return {}, [], None
        params = {"news_id": found["news_id"]}
        async with self._pool.connection() as conn:
            cur = conn.cursor(row_factory=rows.dict_row)
            await cur.execute(_EVIDENCE_SQL, params)
            evidence = await cur.fetchall()
        review = await self._one(_REVIEW_SQL, params)
        return found, evidence, review
